### PR TITLE
feat: add support for .NET SDK V4

### PR DESCRIPTION
**Pull request by 96malhar**
*Issue #, if available:*
DOTNET-7762

*Description of changes:*
This PR updates the .NET SDK dependencies to the latest version 4.0.0-preview.4

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .autover/changes/477dcfa7-055f-4342-8633-de16951d29c1.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "Projects": [
+    {
+      "Name": "AWS.Messaging",
+      "Type": "Major",
+      "ChangelogMessages": [
+        "Updated the .NET SDK dependencies to the latest version 4.0.0-preview.4"
```
Comment (philasmar) [inline .autover/changes/477dcfa7-055f-4342-8633-de16951d29c1.json:7]: Call out Trimming support if you add it

File: src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj [hunk 2]
```diff
@@ -16,7 +16,7 @@
     <WarningsAsErrors>CA1727</WarningsAsErrors>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\..\public.snk</AssemblyOriginatorKeyFile>
-    <Version>0.10.0</Version>
+    <Version>2.0.0-preview.1</Version>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
     <IncludeSymbols>true</IncludeSymbols>
```
Comment (philasmar) [inline src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj:22]: Any reason why we're not enabling Trimming for this library?

File: src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj [hunk 3]
```diff
@@ -16,7 +16,7 @@
     <WarningsAsErrors>CA1727</WarningsAsErrors>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\..\public.snk</AssemblyOriginatorKeyFile>
-    <Version>0.10.0</Version>
+    <Version>2.0.0-preview.1</Version>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
```
Comment (philasmar) [inline src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj:21]: Could you please update `<PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="All" />` to the latest version to match all our other libraries. I think it's 8.0.0.

File: src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj [hunk 4]
```diff
@@ -18,7 +18,7 @@
     <WarningsAsErrors>CA1727</WarningsAsErrors>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\..\public.snk</AssemblyOriginatorKeyFile>
-    <Version>0.9.1</Version>
+    <Version>2.0.0-preview.1</Version>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
```
Comment (philasmar) [inline src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj:23]: Could you please update `<PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="All" />` to the latest version to match all our other libraries. I think it's 8.0.0.

File: src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj [hunk 5]
```diff
@@ -18,7 +18,7 @@
...
     <IncludeSymbols>true</IncludeSymbols>
```
Comment (philasmar) [inline src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj:24]: Any reason why we're not enabling Trimming for this library?

File: src/AWS.Messaging/AWS.Messaging.csproj [hunk 6]
```diff
@@ -17,18 +17,18 @@
     <WarningsAsErrors>CA1727</WarningsAsErrors>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\..\public.snk</AssemblyOriginatorKeyFile>
-    <Version>0.9.2</Version>
+    <Version>2.0.0-preview.1</Version>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
     <IncludeSymbols>true</IncludeSymbols>
     <SymbolPackageFormat>snupkg</SymbolPackageFormat>
```
Comment (philasmar) [inline src/AWS.Messaging/AWS.Messaging.csproj:24]: Any reason why we're not enabling Trimming for this library?

File: src/AWS.Messaging/AWS.Messaging.csproj [hunk 7]
```diff
@@ -17,18 +17,18 @@
     <WarningsAsErrors>CA1727</WarningsAsErrors>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\..\public.snk</AssemblyOriginatorKeyFile>
-    <Version>0.9.2</Version>
+    <Version>2.0.0-preview.1</Version>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
     <IncludeSymbols>true</IncludeSymbols>
```
Comment (philasmar) [inline src/AWS.Messaging/AWS.Messaging.csproj:23]: Could you please update <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="All" /> to the latest version to match all our other libraries. I think it's 8.0.0.

**Review by philasmar — COMMENTED** [review]
Should we add a .NET8 target across all projects and drop .NET6? .NET6 will go out of support next month. By the time we put out the V4 upgrade, .NET6 will have gone out of support.

File: src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj (hunk 2 above)
Comment (normj) [reply src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj:22]: We haven't solved the dynamic reflection problem yet in this library. It won't be a simple switch with the V4 work and should be done separately.

File: .autover/changes/477dcfa7-055f-4342-8633-de16951d29c1.json (hunk 1 above)
Comment (96malhar) [reply .autover/changes/477dcfa7-055f-4342-8633-de16951d29c1.json:7]: We need to move away from reflection based serialization to source generation serialization to support trimming. We have a separate backlog item for this feature.

File: src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj (hunk 2 above)
Comment (96malhar) [reply src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj:22]: We need to move away from reflection based serialization to source generation serialization to support trimming. We have a separate backlog item for this feature.

File: src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj (hunk 5 above)
Comment (96malhar) [reply src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj:24]: We need to move away from reflection based serialization to source generation serialization to support trimming. We have a separate backlog item for this feature.

File: src/AWS.Messaging/AWS.Messaging.csproj (hunk 6 above)
Comment (96malhar) [reply src/AWS.Messaging/AWS.Messaging.csproj:24]: We need to move away from reflection based serialization to source generation serialization to support trimming. We have a separate backlog item for this feature.

File: src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj (hunk 3 above)
Comment (96malhar) [reply src/AWS.Messaging.Lambda/AWS.Messaging.Lambda.csproj:21]: Done

File: src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj (hunk 4 above)
Comment (96malhar) [reply src/AWS.Messaging.Telemetry.OpenTelemetry/AWS.Messaging.Telemetry.OpenTelemetry.csproj:23]: Done

File: src/AWS.Messaging/AWS.Messaging.csproj (hunk 7 above)
Comment (96malhar) [reply src/AWS.Messaging/AWS.Messaging.csproj:23]: Done

**Review by philasmar — APPROVED** [review]

**Review by normj — APPROVED** [review]

**Pull request merged**